File: day03/day03.py
from collections import defaultdict
from itertools import filterfalse

data = [l.strip() for l in open('input.txt').readlines()]
#data = ['00100','11110','10110','10111','10101','01111','00111','11100','10000','11001','00010','01010']

total = 0
ones = defaultdict(lambda: 0)
for line in data:
    total += 1
    for i in range(len(line)):
        ones[i] += int(line[i])
gamma = 0
epsilon = 0
for i in range(len(ones)):
    gamma = gamma << 1
    epsilon = epsilon << 1
    if ones[i] > total/2:
        gamma += 1
    else:
        epsilon += 1
print(f'γ: {gamma:b}={gamma}, ε: {epsilon:b}={epsilon}, γ•ε: {gamma*epsilon}')

def most_common_bit(lines, i):
    n = 0
    for l in lines:
        if l[i] == '1':
            n += 1
    if n >= len(lines)/2:
        return '1'
    return '0'

filtered0 = data
filtered1 = data
for i in range(len(ones)):
    bit0 = most_common_bit(filtered0, i)
    bit1 = most_common_bit(filtered1, i)
    if len(filtered0) > 1:
        filtered0 = [l for l in filtered0 if l[i] == bit0]
    if len(filtered1) > 1:
        filtered1 = [l for l in filtered1 if l[i] != bit1]
    if len(filtered0) == 1 and len(filtered1) == 1:
        break

o2 = int(filtered0[0], 2)
co2 = int(filtered1[0], 2)
print(f'O₂: {o2:b}={o2}, CO₂: {co2:b}={co2}, product: {o2*co2}')

File: day03/day03.rs
use std::cmp::min;
use std::error::Error;
use std::fs::File;
use std::io::BufRead;
use std::io::BufReader;

fn main() -> Result<(), Box<dyn Error>> {
    let f = File::open("input.txt")?;
    let mut reader = BufReader::new(f);
    let data = parse(&mut reader)?;
    let (gamma, epsilon) = gamma_epsilon(&data);
    println!(
        "γ: {:012b}={}, ε: {:012b}={}, γ•ε: {}",
        gamma,
        gamma,
        epsilon,
        epsilon,
        gamma * epsilon
    );

    let o2 = filter(&data, true);
    let co2 = filter(&data, false);
    println!(
        "O₂: {:12b}={}, CO₂: {:12b}={}, product: {}",
        o2,
        o2,
        co2,
        co2,
        o2 * co2
    );

    Ok(())
}

fn parse<R: BufRead>(r: &mut R) -> Result<Data, Box<dyn Error>> {
    let mut n = 16;
    let mut result = Vec::new();
    for line in r.lines() {
        let v = u16::from_str_radix(&line?, 2)?;
        n = min(n, v.leading_zeros());
        result.push(v);
    }
    Ok(Data {
        n: 16 - n as usize,
        nums: result,
    })
}

struct Data {
    n: usize,
    nums: Vec<u16>,
}

fn bit_is_commonly_set(nums: &Vec<u16>, i: usize) -> bool {
    let mut ones = 0;
    let mut zeros = 0;
    for v in nums {
        if (v >> i) & 1 == 0 {
            zeros += 1;
        } else {
            ones += 1;
        }
    }
    return ones >= zeros;
}

fn gamma_epsilon(data: &Data) -> (u64, u64) {
    let mut gamma = 0;
    let mut epsilon = 0;
    for i in (0..data.n).rev() {
        gamma <<= 1;
        epsilon <<= 1;
        if bit_is_commonly_set(&data.nums, i) {
            gamma += 1;
        } else {
            epsilon += 1;
        }
    }
    (gamma, epsilon)
}

fn filter(data: &Data, common: bool) -> u64 {
    let mut filtered = data.nums.to_vec();
    for i in (0..data.n).rev() {
        let want_common = common == bit_is_commonly_set(&filtered, i);
        let mut j = 0;
        for k in 0..filtered.len() {
            let v = filtered[k];
            let is_set = (v >> i) & 1 != 0;
            if is_set == want_common {
                filtered[j] = v;
                j += 1;
            }
        }
        filtered.truncate(j);
        if filtered.len() == 1 {
            break;
        }
    }
    if filtered.len() > 1 {
        panic!("filtered set has more than one element");
    }
    return filtered[0] as u64;
}
